File: backend/routers/kb_router.py
"""
YUMMY Backend - Knowledge Base Router
Endpoints: /kb/*
"""

from fastapi import APIRouter, BackgroundTasks, HTTPException
from config import DB
from dependencies import require_repo
from services.scan_service import run_scan
from services.github_service import github_fetch, github_raw

router = APIRouter(prefix="/kb", tags=["Knowledge Base"])


@router.get("")
def get_knowledge_base():
    """
    Xem knowledge base hiện tại.
    Trả về tree, insights và project summary.
    """
    kb = DB["knowledge_base"]
    return {
        "file_count": len(kb["tree"]),
        "insight_count": len(kb["insights"]),
        "has_summary": bool(kb["project_summary"]),
        "tree": kb["tree"],
        "insights": kb["insights"],
        "project_summary": kb["project_summary"]
    }


@router.post("/scan")
async def start_scan(background_tasks: BackgroundTasks):
    """
    Bắt đầu quét và index codebase từ GitHub.
    
    - Chạy background (non-blocking)
    - Poll GET /kb/scan/status để theo dõi tiến trình
    - Kết quả lưu vào GET /kb
    
    Cần chạy POST /config/setup trước.
    """
    require_repo()

    if DB.get("scan_status") and DB["scan_status"].get("running"):
        raise HTTPException(
            409,
            "Scan đang chạy. Poll GET /kb/scan/status để theo dõi, hoặc đợi hoàn tất."
        )

    background_tasks.add_task(run_scan)
    return {
        "status": "started",
        "message": "Scan đang chạy background. Poll GET /kb/scan/status để theo dõi."
    }


@router.get("/scan/status")
def get_scan_status():
    """
    Kiểm tra tiến trình scan hiện tại.
    
    Response:
    - running: bool
    - text: mô tả bước đang làm
    - progress: 0-100
    - error: true nếu có lỗi
    """
    return DB.get("scan_status") or {
        "running": False,
        "text": "Chưa có scan nào được khởi động.",
        "progress": 0
    }


@router.get("/file")
async def get_file_content(path: str):
    """
    Xem nội dung của một file trong repo (IDE Simulator).
    
    Query param:
        path: Đường dẫn file, ví dụ: src/main.py
    """
    ri = require_repo()

    # Lấy default branch
    repo_resp = await github_fetch(f"/repos/{ri['owner']}/{ri['repo']}")
    if repo_resp.status_code != 200:
        raise HTTPException(502, "Không thể kết nối GitHub API.")
    branch = repo_resp.json().get("default_branch", "main")

    content = await github_raw(ri["owner"], ri["repo"], branch, path)
    return {
        "path": path,
        "content": content,
        "branch": branch,
        "repo": f"{ri['owner']}/{ri['repo']}"
    }


@router.delete("")
def clear_knowledge_base():
    """
    Xóa toàn bộ knowledge base (tree, insights, summary).
    Dùng trước khi scan lại từ đầu.
    """
    DB["knowledge_base"] = {"tree": [], "insights": [], "project_summary": ""}
    DB["scan_status"] = None
    return {"status": "ok", "message": "Knowledge base đã được xóa."}

File: backend-ts/src/routers/kb.router.ts
/**
 * Knowledge Base router — /kb/*.
 * Mirrors backend/routers/kb_router.py.
 */
import { createRoute, OpenAPIHono, z } from '@hono/zod-openapi';
import { type Bindings, createDb } from '../db/client.js';
import { kbRepo } from '../db/repositories/kb.repo.js';
import { scanStatusRepo } from '../db/repositories/scan-status.repo.js';
import { conflict, HttpError } from '../lib/errors.js';
import { requireRepo } from '../lib/guards.js';
import { ErrorSchema } from '../schemas/common.schema.js';
import {
  FileContentSchema,
  FileQuerySchema,
  KnowledgeBaseSchema,
  ScanStatusResponseSchema,
} from '../schemas/kb.schema.js';
import { getRepoInfo, githubRaw } from '../services/github/github.service.js';
import { runScan } from '../services/scan/scan.service.js';

export const kbRouter = new OpenAPIHono<{ Bindings: Bindings }>();

const json = <T extends z.ZodTypeAny>(schema: T) => ({
  'application/json': { schema },
});

// ─── GET /kb ─────────────────────────────────────────────
kbRouter.openapi(
  createRoute({
    method: 'get',
    path: '/kb',
    tags: ['Knowledge Base'],
    responses: {
      200: { content: json(KnowledgeBaseSchema), description: 'KB snapshot' },
    },
  }),
  async (c) => {
    const db = createDb(c.env.DB);
    const snap = await kbRepo.snapshot(db);
    return c.json({
      file_count: snap.tree.length,
      insight_count: snap.insights.length,
      has_summary: !!snap.project_summary,
      tree: snap.tree,
      insights: snap.insights.map((i) => ({ id: i.id, files: i.files, summary: i.summary })),
      project_summary: snap.project_summary,
    });
  },
);

// ─── POST /kb/scan ───────────────────────────────────────
kbRouter.openapi(
  createRoute({
    method: 'post',
    path: '/kb/scan',
    tags: ['Knowledge Base'],
    responses: {
      200: {
        content: json(z.object({ status: z.string(), message: z.string() })),
        description: 'Started',
      },
      400: { content: json(ErrorSchema), description: 'Repo not configured' },
      409: { content: json(ErrorSchema), description: 'Already running' },
    },
  }),
  async (c) => {
    const db = createDb(c.env.DB);
    await requireRepo(db);
    const cur = await scanStatusRepo.get(db);
    if (cur?.running) {
      throw conflict('Scan is already running. Poll GET /kb/scan/status to track progress.');
    }
    // Fire-and-forget background task; matches FastAPI BackgroundTasks behaviour.
    runScan(db).catch((e) => console.error('[scan]', e));
    return c.json(
      {
        status: 'started',
        message: 'Scan started in background. Poll GET /kb/scan/status to track progress.',
      },
      200,
    );
  },
);

// ─── GET /kb/scan/status ─────────────────────────────────
kbRouter.openapi(
  createRoute({
    method: 'get',
    path: '/kb/scan/status',
    tags: ['Knowledge Base'],
    responses: {
      200: { content: json(ScanStatusResponseSchema), description: 'Status' },
    },
  }),
  async (c) => {
    const db = createDb(c.env.DB);
    const s = await scanStatusRepo.get(db);
    if (!s) {
      return c.json({
        running: false,
        text: 'No scan has been started yet.',
        progress: 0,
        error: false,
      });
    }
    return c.json(s);
  },
);

// ─── GET /kb/file?path=... ───────────────────────────────
kbRouter.openapi(
  createRoute({
    method: 'get',
    path: '/kb/file',
    tags: ['Knowledge Base'],
    request: { query: FileQuerySchema },
    responses: {
      200: { content: json(FileContentSchema), description: 'File content' },
      400: { content: json(ErrorSchema), description: 'Repo not configured' },
      404: { content: json(ErrorSchema), description: 'File not found' },
      502: { content: json(ErrorSchema), description: 'GitHub error' },
    },
  }),
  async (c) => {
    const db = createDb(c.env.DB);
    const { path } = c.req.valid('query');
    const ri = await requireRepo(db);

    let branch: string;
    try {
      const repoData = await getRepoInfo(db, ri.owner, ri.repo);
      branch = repoData.default_branch ?? 'main';
    } catch {
      throw new HttpError(502, 'Unable to connect to GitHub API.');
    }

    const content = await githubRaw(db, ri.owner, ri.repo, branch, path);
    return c.json({ path, content, branch, repo: `${ri.owner}/${ri.repo}` }, 200);
  },
);

// ─── DELETE /kb ──────────────────────────────────────────
kbRouter.openapi(
  createRoute({
    method: 'delete',
    path: '/kb',
    tags: ['Knowledge Base'],
    responses: {
      200: {
        content: json(z.object({ status: z.string(), message: z.string() })),
        description: 'Cleared',
      },
    },
  }),
  async (c) => {
    const db = createDb(c.env.DB);
    await kbRepo.resetAll(db);
    await scanStatusRepo.clear(db);
    return c.json({ status: 'ok', message: 'Knowledge base cleared.' });
  },
);
